### ISSUE TITLE
line 153, in __update_resource_assignments\nAttributeError: 'NoneType' object has no attribute 'patch'\n"

### ISSUE THREAD

**Issue by bigmagic** [HewlettPackard/oneview-python#98]
Hi,

Versions used:
- Python: 3.8
- Ansible: 2.9.13
- hpeOneView: 5.4.0
- oneview-ansible : latest version
- Operating System: RHEL 8.2
- OneView: 5.20.00-0419867

I am using OneView Ansible module on Synergy and I am facing a bug when I am trying to assign a resource to a scope.

Python returns a NoneType error:

"/tmp/ansible_oneview_scope_payload_d5cf8bie/ansible_oneview_scope_payload.zip/ansible/modules/oneview_scope.py\", line 153, in __update_resource_assignments\nAttributeError: 'NoneType' object has no attribute 'patch'\n", "module_stdout": "", "msg": "MODULE FAILURE\

The self.current_resource objet is not defined at line 153 of the source code:
https://github.com/HewlettPackard/oneview-ansible/blob/master/library/oneview_scope.py

You will find below my playbook:

```
---
- hosts: oneview
  vars:
    # Python
    ansible_python_interpreter: /usr/bin/python3.8
    config: "/root/oneview.json"
    uri_scope: "/rest/server-profile-templates/bdbba95d-3272-4eed-8a72-471b8e90120a"
  vars_files:
    - /root/oneview_config.yml

  tasks:
    - name: Get SPT
      oneview_server_profile_template_facts:
        config: "{{ config }}"
        name: "ProfileTemplate102"
      delegate_to: localhost
      register: spt

    - debug: var=spt["ansible_facts"]["server_profile_templates"][0]["uri"]

    - name: Update the scope resource assignments, adding a resource
      oneview_scope:
        config: '{{ config }}'
        state: resource_assignments_updated
        data:
          name: "ScopeDev"
          resourceAssignments:
            addedResourceUris: '{{ spt["ansible_facts"]["server_profile_templates"][0]["uri"] }}'
      delegate_to: localhost
```

The oneview.json file contains:

```
[root@rhel82 ~]# cat oneview.json
{
  "ip": "192.168.10.55",
  "credentials": {
    "userName": "Administrator",
    "authLoginDomain": "",
    "password": "****"
  },
  "api_version": 1600
}
```
I thank you,

Best Regards,
Nicolas Portais

Below the execution of the playbook:

```
# ansible-playbook -v assign_scope.yml
Using /etc/ansible/ansible.cfg as config file

PLAY [oneview] ***************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************************************************************************************************
ok: [192.168.10.222]

TASK [Get SPT] ***************************************************************************************************************************************************************************************************
ok: [192.168.10.222] => {"ansible_facts": {"server_profile_templates": [{"affinity": "Bay", "bios": {"complianceControl": "Unchecked", "manageBios": false, "overriddenSettings": []}, "boot": {"complianceControl": "Unchecked", "manageBoot": false, "order": []}, "bootMode": {"complianceControl": "Unchecked", "manageMode": false, "mode": null, "pxeBootPolicy": null, "secureBoot": "Unmanaged"}, "category": "server-profile-templates", "connectionSettings": {"complianceControl": "CheckedMinimum", "connections": [{"functionType": "Ethernet", "id": 1, "isolatedTrunk": false, "lagName": null, "managed": true, "name": "Prod-NetworkSet-1", "networkName": null, "networkUri": "/rest/ethernet-networks/3f794477-d9f6-45ea-920b-0a18b9c87151", "portId": "Mezz 3:1-c", "requestedMbps": "2500", "requestedVFs": "Auto"}], "manageConnections": true}, "created": "2020-10-01T08:04:39.129Z", "description": "", "eTag": "1601541803405/5", "enclosureGroupUri": "/rest/enclosure-groups/aade80e1-906e-4198-b803-970eb4058de9", "firmware": {"complianceControl": "Unchecked", "forceInstallFirmware": false, "manageFirmware": false}, "hideUnusedFlexNics": true, "iscsiInitiatorNameType": "AutoGenerated", "localStorage": {"complianceControl": "Unchecked", "controllers": [], "sasLogicalJBODs": []}, "macType": "Virtual", "managementProcessor": {"complianceControl": "Unchecked", "manageMp": false, "mpSettings": []}, "modified": "2020-10-01T08:43:23.405Z", "name": "ProfileTemplate102", "osDeploymentSettings": null, "sanStorage": {"complianceControl": "Unchecked", "manageSanStorage": false, "sanSystemCredentials": [], "volumeAttachments": []}, "scopesUri": "/rest/scopes/resources/rest/server-profile-templates/bdbba95d-3272-4eed-8a72-471b8e90120a", "serialNumberType": "Virtual", "serverHardwareTypeUri": "/rest/server-hardware-types/BC459A8E-1818-421F-98F7-BA4C5CCBF840", "serverProfileDescription": "", "state": null, "status": "OK", "type": "ServerProfileTemplateV8", "uri": "/rest/server-profile-templates/bdbba95d-3272-4eed-8a72-471b8e90120a", "wwnType": "Virtual"}]}, "changed": false}

TASK [debug] *****************************************************************************************************************************************************************************************************
ok: [192.168.10.222] => {
    "spt[\"ansible_facts\"][\"server_profile_templates\"][0][\"uri\"]": "/rest/server-profile-templates/bdbba95d-3272-4eed-8a72-471b8e90120a"
}

TASK [Update the scope resource assignments, adding a resource] **************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'NoneType' object has no attribute 'patch'
fatal: [192.168.10.222]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1601989582.9783294-31067-40376887274795/AnsiballZ_oneview_scope.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1601989582.9783294-31067-40376887274795/AnsiballZ_oneview_scope.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1601989582.9783294-31067-40376887274795/AnsiballZ_oneview_scope.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible.modules.oneview_scope', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.8/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.8/runpy.py\", line 95, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib64/python3.8/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_oneview_scope_payload_tuf7y14q/ansible_oneview_scope_payload.zip/ansible/modules/oneview_scope.py\", line 184, in <module>\n  File \"/tmp/ansible_oneview_scope_payload_tuf7y14q/ansible_oneview_scope_payload.zip/ansible/modules/oneview_scope.py\", line 180, in main\n  File \"/tmp/ansible_oneview_scope_payload_tuf7y14q/ansible_oneview_scope_payload.zip/ansible/module_utils/oneview.py\", line 475, in run\n  File \"/tmp/ansible_oneview_scope_payload_tuf7y14q/ansible_oneview_scope_payload.zip/ansible/modules/oneview_scope.py\", line 145, in execute_module\n  File \"/tmp/ansible_oneview_scope_payload_tuf7y14q/ansible_oneview_scope_payload.zip/ansible/modules/oneview_scope.py\", line 154, in __update_resource_assignments\nAttributeError: 'NoneType' object has no attribute 'patch'\n", "module_stdout": "#### self.current_resource <__main__.ScopeModule object at 0x7f20eee53400>\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}

PLAY RECAP *******************************************************************************************************************************************************************************************************
192.168.10.222             : ok=3    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by AsisBagga** [comment]
@bigmagic Hi Nicolas, I have tried to reproduce the error and its occurs when the specified scope in the task is not valid resources name. 
Can confirm if scope: "ScopeDev" is already present in OneView? 
If No, then please try to create the scope on the specified name and then try to the update operation for resourceAssignments.
If Yes, can you gather the facts for [ScopeDev ](https://github.com/HewlettPackard/oneview-ansible/blob/master/examples/oneview_scope_facts.yml)and [Server Profile](https://github.com/HewlettPackard/oneview-ansible/blob/master/examples/oneview_server_profile_template_facts.yml) and share it with us please?

**Comment by AsisBagga** [comment]
@bigmagic  Hi Nicolas, I have tried to reproduce the error and one of the reasons why it occurs is when the specified scope in the task is not valid resources name or not present in oneview appliance. 
Can confirm if scope: "ScopeDev" is already present in OneView? 
If No, then please try to create the scope of the specified name i.e. "ScopeDev" and then try to the update operation for resourceAssignments.
If Yes, can you gather the facts for [ScopeDev ](https://github.com/HewlettPackard/oneview-ansible/blob/master/examples/oneview_scope_facts.yml)and [Server Profile](https://github.com/HewlettPackard/oneview-ansible/blob/master/examples/oneview_server_profile_template_facts.yml) and share it with us please?

**Comment by bigmagic** [comment]
Hi Asis,

Yes, it seems that I have interverted "DevScope" and "ScopeDev" but it should not failed with a stack trace. The module should returns an error message saying that the scope doesn't exist.

Below the scopes defined into my environment:

![image](https://user-images.githubusercontent.com/7739748/95336301-cce8b080-08b0-11eb-9b86-939054d95b7d.png)

I have corrected the typo mistake (sorry for that) but we also have an idempotence issue when we execute the playbook for the second times : the scope assignment should not be modified when the assignment already exists.

Below the corrected test result:

```
$ docker run -it --rm -v playbooks:/playbooks hewlettpackardenterprise/hpe-oneview-sdk-for-ansible:v5.7.0-OV5.3 /bin/bash
root@ce16fe72f3a6:~/oneview-ansible# cd /playbooks/
root@ce16fe72f3a6:/playbooks# ansible-playbook  -i localhost assign_scope.yml
 [WARNING]: Unable to parse /playbooks/localhost as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] ******************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************
ok: [localhost]

TASK [Get SPT] ********************************************************************************************************************************************************************************
ok: [localhost -> localhost]

TASK [debug] **********************************************************************************************************************************************************************************
ok: [localhost] => {
    "spt[\"ansible_facts\"][\"server_profile_templates\"][0][\"uri\"]": "/rest/server-profile-templates/bdbba95d-3272-4eed-8a72-471b8e90120a"
}

TASK [Update the scope resource assignments, adding a resource] *******************************************************************************************************************************
changed: [localhost -> localhost]

PLAY RECAP ************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=1    unreachable=0    failed=0
```

I have improved my example and search the scope by name : 

```
root@ce16fe72f3a6:/playbooks# cat assign_scope2.yml
---
- hosts: localhost
  vars:
    # Python
    ansible_python_interpreter: /usr/local/bin/python3.6
    config: "/playbooks/oneview.json"
    uri_scope: "/rest/server-profile-templates/bdbba95d-3272-4eed-8a72-471b8e90120a"
  vars_files:
    - /playbooks/oneview_config.yml

  tasks:

    - name: Gather scope by name
      oneview_scope_facts:
        config: "{{ config }}"
        name: "Scope2"
      delegate_to: localhost
      register: scope_name

    - debug: var=scope_name

    - name: Get SPT
      oneview_server_profile_template_facts:
        config: "{{ config }}"
        name: "ProfileTemplate102"
      delegate_to: localhost
      register: spt

    - debug: var=spt["ansible_facts"]["server_profile_templates"][0]["uri"]

    - name: Update the scope resource assignments, adding a resource
      oneview_scope:
        config: '{{ config }}'
        state: resource_assignments_updated
        data:
          name: '{{ scope_name["ansible_facts"]["scopes"][0]["name"] }}'
          resourceAssignments:
            addedResourceUris: '{{ spt["ansible_facts"]["server_profile_templates"][0]["uri"] }}'
      delegate_to: localhost
```

And it failed : 

```
root@ce16fe72f3a6:/playbooks# ansible-playbook  -i localhost assign_scope2.yml
 [WARNING]: Unable to parse /playbooks/localhost as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] ******************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************
ok: [localhost]

TASK [Gather scope by name] *******************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: Value of unknown type: <class 'hpOneView.resources.settings.scopes.Scopes'>, <hpOneView.resources.settings.scopes.Scopes object at 0x7fea560a2eb8>
fatal: [localhost -> localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1602078220.7700422-130314937329244/AnsiballZ_oneview_scope_facts.py\", line 113, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1602078220.7700422-130314937329244/AnsiballZ_oneview_scope_facts.py\", line 105, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1602078220.7700422-130314937329244/AnsiballZ_oneview_scope_facts.py\", line 48, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/usr/local/lib/python3.6/imp.py\", line 235, in load_module\n    return load_source(name, filename, file)\n  File \"/usr/local/lib/python3.6/imp.py\", line 170, in load_source\n    module = _exec(spec, sys.modules[name])\n  File \"<frozen importlib._bootstrap>\", line 618, in _exec\n  File \"<frozen importlib._bootstrap_external>\", line 678, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n  File \"/tmp/ansible_oneview_scope_facts_payload_yhe829ak/__main__.py\", line 125, in <module>\n  File \"/tmp/ansible_oneview_scope_facts_payload_yhe829ak/__main__.py\", line 121, in main\n  File \"/tmp/ansible_oneview_scope_facts_payload_yhe829ak/ansible_oneview_scope_facts_payload.zip/ansible/module_utils/oneview.py\", line 780, in run\n  File \"/tmp/ansible_oneview_scope_facts_payload_yhe829ak/ansible_oneview_scope_facts_payload.zip/ansible/module_utils/basic.py\", line 2348, in exit_json\n  File \"/tmp/ansible_oneview_scope_facts_payload_yhe829ak/ansible_oneview_scope_facts_payload.zip/ansible/module_utils/basic.py\", line 2341, in _return_formatted\n  File \"/tmp/ansible_oneview_scope_facts_payload_yhe829ak/ansible_oneview_scope_facts_payload.zip/ansible/module_utils/basic.py\", line 526, in remove_values\n  File \"/tmp/ansible_oneview_scope_facts_payload_yhe829ak/ansible_oneview_scope_facts_payload.zip/ansible/module_utils/basic.py\", line 505, in _remove_values_conditions\nTypeError: Value of unknown type: <class 'hpOneView.resources.settings.scopes.Scopes'>, <hpOneView.resources.settings.scopes.Scopes object at 0x7fea560a2eb8>\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
        to retry, use: --limit @/playbooks/assign_scope2.retry

PLAY RECAP ************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

Best Regards,
Nicolas Portais

**Comment by AsisBagga** [comment]
@bigmagic Hi Nicolas, Thanks for pointing this out, I was able to reproduce the issue today, ill update you once it's fixed.

**Comment by AsisBagga** [comment]
@bigmagic Hi Nicolas, 
The request to add/delete the resource uri to an existing scope is a patch request, hence is not idempotent.
This issue for not getting the scope by its name has been addressed on [PR #584](https://github.com/HewlettPackard/oneview-ansible/pull/584) and will reflect in our coming releases, for now i can suggest as a work around you may fetch the scope facts like this
```
- name: Gather paginated, filtered and sorted facts about Scopes
   oneview_scope_facts:
     config: "{{ config }}"
      params:
       query: name eq 'SampleScope'
   delegate_to: localhost
   register: scope_name
```
Also it has been taken care that if a scope does not exist during an update operation, it will return an error instead of the stack trace.

**Comment by bigmagic** [comment]
Hi Asis, Thanks for these details and the PR. We will use this workaround.
